src/services/Batcher.ts

```diff
@@ -0,0 +1,102 @@
+import { MaybePromise } from '@/types'
+
+export type BatchCallbackLookup<V, R> = (value: V) => MaybePromise<R>
+export type BatchCallback<V, R> = (values: V[]) => MaybePromise<Map<V, R>> | BatchCallbackLookup<V, R>
+export type BatchOptions = {
+  maxBatchSize?: number,
+  maxCycles?: number,
+}
+
+type Batch<V, R> = Map<V, BatchPromise<R>>
+type BatchPromise<T> = {
+  resolve: (value: T) => void,
+  reject: (reason?: unknown) => void,
+}
+
+export class Batcher<V, R> {
+  private readonly queue: Batch<V, R> = new Map()
+  private timeout: ReturnType<typeof setTimeout> | null = null
+  private readonly callback: BatchCallback<V, R>
+  private readonly options: BatchOptions
+  private readonly loops: number = 0
+
+  public constructor(callback: BatchCallback<V, R>, options: BatchOptions = {}) {
+    this.callback = callback
+    this.options = options
+  }
+
+  public batch(value: V): Promise<R> {
+
+    const response = new Promise<R>((resolve, reject) => {
+      this.queue.set(value, { resolve, reject })
+    })
+
+    this.requestBatchProcessOnNextCycle()
+
+    return response
+  }
+
+  private requestBatchProcessOnNextCycle(): void {
+    if (this.timeout) {
+      clearTimeout(this.timeout)
+    }
+
+    const maxBatchSize = this.options.maxBatchSize ?? Infinity
```

Comment:
```suggestion
    const { maxBatchSize = Infinity } = this.options
```
I'm surprised eslint isn't warning about this. Maybe the coalesce overrides the rule